XenAdmin/Dialogs/AddVGPUDialog.cs

```diff
@@ -64,51 +67,48 @@ private void PopulateComboBox()
                 {
                     var enabledTypes = _vm.Connection.ResolveAll(gpu_group.enabled_VGPU_types);
                     var allTypes = _vm.Connection.ResolveAll(gpu_group.supported_VGPU_types);
-
                     var disabledTypes = allTypes.FindAll(t => !enabledTypes.Exists(e => e.opaque_ref == t.opaque_ref));
 
-                    if (gpu_group.HasVGpu())
-                    {
-                        allTypes.Sort();
-                        allTypes.Reverse();
-                        comboBoxTypes.Items.Add(new GpuTuple(gpu_group, allTypes.ToArray())); // Group item
-                    }
+                    allTypes.Sort();
+                    allTypes.Reverse();
 
-                    List<string> commonTypes = new List<string>();
-                    foreach (VGPU_type vgpuType in allTypes)
-                        commonTypes.Add(vgpuType.model_name);
-                    
-                        
-                    foreach (VGPU eVgpu in existingVGpus)
+                    var commonTypes = new List<VGPU_type>();
+                    commonTypes.AddRange(allTypes);
+
+                    foreach (var eVgpu in existingVGpus)
                     {
-                        VGPU_type etype = _vm.Connection.Resolve(eVgpu.type);
-                        List<string> tmpCommonTypes = new List<string>();
-                        foreach (string compatible_type in etype.compatible_types_in_vm)
-                            if (commonTypes.Contains(compatible_type))
-                                tmpCommonTypes.Add(compatible_type);
-                        commonTypes = tmpCommonTypes;
+                        var etype = _vm.Connection.Resolve(eVgpu.type);
+                        foreach (var vgpuType in allTypes)
+                        {
+                            if (!etype.compatible_types_in_vm.Contains(vgpuType.model_name))
+                                commonTypes.Remove(vgpuType);
+                        }
                     }
-                    foreach (var vgpuType in allTypes)
-                        if (commonTypes.Contains(vgpuType.model_name))
-                            comboBoxTypes.Items.Add(new GpuTuple(gpu_group, vgpuType, disabledTypes.ToArray())); // GPU_type item
 
+                    if (gpu_group.HasVGpu() && commonTypes.Count > 0)
+                        comboBoxTypes.Items.Add(new GpuTuple(gpu_group, allTypes.ToArray())); // Group item
+
+                    foreach (var vgpuType in commonTypes)
+                        comboBoxTypes.Items.Add(new GpuTuple(gpu_group, vgpuType, disabledTypes.ToArray())); // GPU_type item
+                    
                 }
             }
-            comboBoxTypes.SelectedItem = noneItem;
+            if (comboBoxTypes.Items.Count == 0)

```

Comment:
We should never be in the situation where the combobox is empty. But if this ever occurs, I think an empty, enabled combobox looks better than a disabled one, especially because it is the only control on this dialog

Reply:
Agree, thanks.